### PR TITLE
[IMP] web_editor, website: allow to create a link to a document

### PR DESCRIPTION
**Pull request by SergeBayet**
This commit provides a new button next to the URL input in the "link
tools" to easily create links to documents
It enhances user experience by allowing selection from the media library
instead of manually entering a URL.
A test has been made.

task-4100016

### CODE REVIEW

**Comment by Julien00859** [comment]
Almost a double win, miss the 4100000th task by 16 :joy:

**Comment by SergeBayet** [comment]
> We have a winner! This is the 177777th pull request :D Congrat!

Holy crap!! :) made my day. :)

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js [hunk 1]
```diff
@@ -425,6 +444,18 @@ export class LinkTools extends Link {
     _onClickCheckbox(ev) {
         const $target = $(ev.target);
         $target.closest('we-button.o_we_checkbox_wrapper').toggleClass('active');
+        if ($target[0].getAttribute("name") === "direct_download") {
+            const el = $target.closest("we-button.o_we_checkbox_wrapper")[0];
```
Comment (detrouxdev) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:448]: can probably just use ev.target for everthing to avoid referencing jQuery, except if you see a difference in behaviour?
```suggestion
            const el = ev.target.closest("we-button.o_we_checkbox_wrapper");
```

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js [hunk 2]
```diff
@@ -481,12 +512,24 @@ export class LinkTools extends Link {
             this.props.wysiwyg.odooEditor.historyStep();
         }
     }
+    _checkDocumentState() {
+        this.state.isDocument = false;
+        this.state.directDownload = true;
```
Comment (detrouxdev) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:517]: Nitpicking but it's odd to me that this would be set to true if it's not a document. I realise that it doesn't really matter and it probably makes more sense since the media dialog returns a "&download=1" link

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js [hunk 3]
```diff
@@ -425,6 +444,18 @@ export class LinkTools extends Link {
     _onClickCheckbox(ev) {
         const $target = $(ev.target);
         $target.closest('we-button.o_we_checkbox_wrapper').toggleClass('active');
+        if ($target[0].getAttribute("name") === "direct_download") {
+            const el = $target.closest("we-button.o_we_checkbox_wrapper")[0];
+            const urlInputEl = this.$el[0].querySelector("#o_link_dialog_url_input");
+            let href = urlInputEl.value;
+            if (el.classList.contains("active")) {
+                href = href.replace("&download=true", "") + "&download=true";
```
Comment (detrouxdev) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:452]: Why is this necessary?

File: addons/web_editor/static/src/js/wysiwyg/widgets/link.js [hunk 4]
```diff
@@ -298,6 +298,17 @@ export class Link extends Component {
...
+        let isDocument = false;
+        let directDownload = true;
+        try {
+            const parsedUrl = new URL(this.props.link.href);
+            
+            if (parsedUrl && parsedUrl.pathname.startsWith("/web/content/")) {
+                isDocument = true;
+                directDownload = parsedUrl.searchParams.get("download") === "true";
+            }
+        } catch {}
```
Comment (detrouxdev) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link.js:310]: Empty try catch triggers a warning in my editor config, so it's probably incorrect (I know I said it'd probably be fine but I guess not) 

I actually think that we should strip the domain from the URL that the media dialog returns, it would remove the "auto-convert relative link" since it would be relative from the start, and it would allow use to just do `this.props.link.href && this.props.link.href.startsWith("/web/content/")` without having to parse the URL. What do you think?

**Review by detrouxdev — CHANGES_REQUESTED** [review]
@SergeBayet Good job, just a quick review with a few remarks and questions :)

@qsm-odoo Should this code somehow also land in the new html_editor ? Since link.js was also part of the old wysiwyg, I assume they converted it somehow? Or should we assume that when working towards feature parity, it will be re-implemented in the new files?

**Comment by qsm-odoo** [comment]
@SergeBayet @detrouxdev 

> Should this code somehow also land in the new html_editor

Depends: if it takes you 10 minutes to do it: yes. Otherwise, no, it can just be for the website (and I guess mass mailing?) builder. Just make sure to add a test so that the feature is not lost when we do use html_editor into website :pray:

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js (hunk 3 above)
Comment (SergeBayet) [reply addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:452]: I wanted to be sure that "&download=true" was not appended manually by the user. But I've refactored.

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js (hunk 2 above)
Comment (SergeBayet) [reply addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:517]: Yes, by default the link returned is appended with &download=true.

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js (hunk 3 above)
Comment (detrouxdev) [reply addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js]: Isn't the fact that the state changes automatically changes the state of the checkbox? I just tried manually adding &download=true and the checkbox ticked itself, so I don't think it should be a concern, you can just assume that if the checkbox is ticked, it is present right? and if not, it is not

Anyway I'm okay with the way you did in the refactor if you think it's worth it to avoid buggy links.

File: addons/website/tests/test_attachment.py [hunk 5]
```diff
@@ -41,3 +42,14 @@ def test_01_type_url_301_image(self):
 
     def test_02_image_quality(self):
         self.start_tour(self.env['website'].get_client_action_url('/'), 'website_image_quality', login="admin")
+
+    @mute_logger('odoo.addons.http_routing.models.ir_http', 'odoo.http')
```
Comment (detrouxdev) [inline addons/website/tests/test_attachment.py:46]: Why do you need to mute the logger here?

Comment (SergeBayet) [reply addons/website/tests/test_attachment.py:46]: > Why do you need to mute the logger here?

@detrouxdev I was inspired by https://github.com/odoo/odoo/commit/d7707a4a88b011686b5531edf54b08dc1a4952c7#diff-ab4a32b8da6461c750e479d0ddc235dc5c388110f80beae5f9d27cadcf34899c

Do you think it's unnecessary?

Comment (detrouxdev) [reply addons/website/tests/test_attachment.py:46]: I think if we don't know the actual reason, muting the logger might hide a sneaky error and we should probably avoid it, but if we have a valid reason it's okay. It's unfortunate that there's no comment on why the logger is muted for that other test.

Comment (SergeBayet) [reply addons/website/tests/test_attachment.py:46]: @bso-odoo Since you're the author of the test I was inspired by. Could you explain why you need to mute the logger in this case? (See the link above). Thanks.

Comment (bso-odoo) [reply addons/website/tests/test_attachment.py:46]: I'd say it was a mistake - and most likely copied from another test v_v

Comment (SergeBayet) [reply addons/website/tests/test_attachment.py:46]: I'll get rid of it ...

**Comment by SergeBayet** [comment]
@detrouxdev It's in order. Something else about this PR?

**Comment by detrouxdev** [comment]
@SergeBayet I'll wait for the tests results without the logger muted just to be sure, then I can R+

**Review by detrouxdev — APPROVED** [review]
Good work, will be quite useful 👍🏻 

@robodoo r+
